File: server/src/utils/jobStore.js
/**
 * @file jobStore.js
 * @description Prompt 1, Part 3/5/6/7/14 — generic Redis-backed job
 * abstraction replacing the in-memory `Map`-based job stores that
 * disappeared on every restart/redeploy (the actual production problem this
 * phase exists to fix).
 *
 * Design
 * ------
 * - Keyed `job:{type}:{jobId}` — `type` namespaces the four call sites that
 *   previously had their own ad-hoc Map (decisionService.js's _BATCH_JOBS,
 *   the ai-service's _EXPLAIN_JOBS/_DECISION_JOBS — this file is the Node
 *   side; app/core/job_store.py is its Python mirror for those two).
 * - Every job carries `organizationId`. getJob() enforces ownership when
 *   called with one (Part 14): a mismatch returns null, identical to "not
 *   found" — confirming *that a job ID exists but belongs to someone else*
 *   is itself a cross-tenant information leak, so this never distinguishes
 *   403 from 404 to the caller.
 * - Crash/restart recovery (Part 6) via a per-process "epoch": each boot of
 *   this process gets a random id. A job is stamped with the epoch of the
 *   process that created it. If a read finds status "running" tagged with a
 *   DIFFERENT epoch than the current process's, the worker that was
 *   computing it is provably gone (this process restarted) — the job is
 *   corrected to "failed" with a clear message instead of being left
 *   "running" forever or silently reported as complete.
 * - TTL on every write (Part 5) so nothing accumulates unbounded — same
 *   30-minute default the old in-memory stores used.
 * - Local-dev fallback: when Redis isn't configured (env.redis.configured
 *   is false — see redisClient.js), this degrades to an in-process Map with
 *   the same TTL behavior. That fallback is intentionally NOT used in
 *   production: config/env.js's production checks plus this session's
 *   Prompt 1 report both call out that Upstash credentials are required
 *   there. Development is allowed to run without Redis so a fresh clone
 *   still boots.
 * - Large results (Part 5 — "do not store unnecessarily large AI results
 *   directly in Redis"): this store still holds the ai-service's batch
 *   result once per job (same as the in-memory version it replaces), scoped
 *   by a short TTL, because it is the only handoff channel between the two
 *   services for that data today. See docs/AUDIT_PROMPT0.md and the Prompt
 *   1 final report's "Remaining Risks" for why explain-batch (unlike
 *   decision-batch, already chunked to 300) is the one job type where this
 *   could matter at full workforce scale, and why fixing that is deferred.
 */

import { randomUUID } from 'crypto';
import { getRedisClient, isRedisConfigured } from './redisClient.js';

const DEFAULT_TTL_SECONDS = 30 * 60;

// Unique per process boot. Used to detect "this running job's process is
// gone" on read — see the module docstring.
export const PROCESS_EPOCH = randomUUID();

const memoryStore = new Map(); // dev-only fallback; see module docstring.
const memoryTimers = new Map();

function redisKey(type, jobId) {
  return `job:${type}:${jobId}`;
}

async function writeRaw(type, jobId, value, ttlSeconds) {
  const k = redisKey(type, jobId);
  if (isRedisConfigured()) {
    await getRedisClient().set(k, value, { ex: ttlSeconds });
    return;
  }
  memoryStore.set(k, value);
  clearTimeout(memoryTimers.get(k));
  const timer = setTimeout(() => memoryStore.delete(k), ttlSeconds * 1000);
  timer.unref?.();
  memoryTimers.set(k, timer);
}

/** @returns {true} if this call created the key, {false} if it already existed (SETNX semantics — Part 7 idempotency). */
async function writeIfAbsent(type, jobId, value, ttlSeconds) {
  const k = redisKey(type, jobId);
  if (isRedisConfigured()) {
    const result = await getRedisClient().set(k, value, { ex: ttlSeconds, nx: true });
    return result !== null;
  }
  if (memoryStore.has(k)) return false;
  memoryStore.set(k, value);
  const timer = setTimeout(() => memoryStore.delete(k), ttlSeconds * 1000);
  timer.unref?.();
  memoryTimers.set(k, timer);
  return true;
}

async function readRaw(type, jobId) {
  const k = redisKey(type, jobId);
  if (isRedisConfigured()) {
    const value = await getRedisClient().get(k);
    return value ?? null;
  }
  return memoryStore.get(k) ?? null;
}

/** Part 6 — corrects a job left "running" by a process that no longer exists. Best-effort persists the correction. */
function recoverIfStale(type, jobId, job) {
  if (job.status === 'running' && job.epoch !== PROCESS_EPOCH) {
    const corrected = {
      ...job,
      status: 'failed',
      error: 'This job was still running when the server restarted and could not be resumed. Please retry.',
      recovered: true,
      updatedAt: new Date().toISOString(),
    };
    writeRaw(type, jobId, corrected, DEFAULT_TTL_SECONDS).catch(() => {});
    return corrected;
  }
  return job;
}

/**
 * Creates a new job with a random ID.
 * @param {string} type - namespace, e.g. 'decision-batch'.
 * @param {object} opts
 * @param {string|null} opts.organizationId - the AUTHENTICATED caller's org
 *   (req.auth.organizationId) — never a client-supplied value (Part 10).
 * @param {number} [opts.ttlSeconds]
 * @returns {Promise<string>} jobId
 */
export async function createJob(type, { organizationId, ttlSeconds = DEFAULT_TTL_SECONDS } = {}) {
  const jobId = randomUUID();
  const now = new Date().toISOString();
  const job = {
    status: 'running',
    organizationId: organizationId ?? null,
    epoch: PROCESS_EPOCH,
    createdAt: now,
    updatedAt: now,
  };
  await writeRaw(type, jobId, job, ttlSeconds);
  return jobId;
}

/**
 * Creates a job under a CALLER-SUPPLIED id only if one doesn't already
 * exist — Part 7's idempotency-key mechanism. Use when the same logical
 * operation (e.g. "batch decisions for org X's ACTIVE employees") must not
 * be started twice concurrently; derive `jobId` deterministically from the
 * operation's own parameters at the call site.
 * @returns {Promise<{jobId: string, created: boolean}>} created=false means
 *   an equivalent job is already in flight (or recently finished) — the
 *   caller should return the existing jobId rather than starting new work.
 */
export async function createJobIdempotent(type, jobId, { organizationId, ttlSeconds = DEFAULT_TTL_SECONDS } = {}) {
  const now = new Date().toISOString();
  const job = {
    status: 'running',
    organizationId: organizationId ?? null,
    epoch: PROCESS_EPOCH,
    createdAt: now,
    updatedAt: now,
  };
  const created = await writeIfAbsent(type, jobId, job, ttlSeconds);
  return { jobId, created };
}

/**
 * @param {object} [opts]
 * @param {string} [opts.organizationId] - if given, enforces ownership;
 *   a mismatch returns null (indistinguishable from "not found" — Part 14).
 */
export async function getJob(type, jobId, { organizationId } = {}) {
  const job = await readRaw(type, jobId);
  if (!job) return null;
  if (organizationId && job.organizationId && String(job.organizationId) !== String(organizationId)) {
    return null;
  }
  return recoverIfStale(type, jobId, job);
}

/** Merges `patch` into the existing job and refreshes updatedAt/TTL. No-op (returns null) if the job doesn't exist (e.g. it already expired). */
export async function updateJob(type, jobId, patch, { ttlSeconds = DEFAULT_TTL_SECONDS } = {}) {
  const existing = await readRaw(type, jobId);
  if (!existing) return null;
  const updated = { ...existing, ...patch, updatedAt: new Date().toISOString() };
  await writeRaw(type, jobId, updated, ttlSeconds);
  return updated;
}

File: ai-service/app/utils/job_store.py
"""
job_store.py
============
Prompt 1, Part 3/5/6/7 — Redis-backed job store replacing the in-memory
dicts in app/api/explain_routes.py (_EXPLAIN_JOBS) and
app/api/decision_routes.py (_DECISION_JOBS). See server/src/utils/jobStore.js
for the Node mirror and the shared design rationale (keying, TTL, crash
recovery via a per-process epoch, why large batch results still pass
through here for now). The two are intentionally structurally identical so
the same mental model applies on either side of the Express<->ai-service
boundary.

Ownership note: these two job types are never polled directly by a browser
— Express is the only caller (shared AI_SERVICE_TOKEN), and it always
supplies the jobId it just created. There is therefore no per-tenant
`organizationId` to enforce here (Express enforces that one layer up, on
its own _BATCH_JOBS-equivalent and on the Decision/Explanation documents
those results get written into). This store still accepts an optional
`owner` tag for defense in depth / future direct exposure, but no current
call site needs it.
"""

from __future__ import annotations

import asyncio
import json
import time
import uuid
from typing import Any, Optional

from app.utils.redis_client import get_redis_client, is_redis_configured

DEFAULT_TTL_SECONDS = 30 * 60

# Unique per process boot — a job stamped with a DIFFERENT epoch than the
# current process's, while still "running", was being computed by a worker
# that no longer exists (this process restarted). See recover_if_stale().
PROCESS_EPOCH = str(uuid.uuid4())

_memory_store: dict[str, dict[str, Any]] = {}
_memory_expiry: dict[str, float] = {}


def _key(job_type: str, job_id: str) -> str:
    return f"job:{job_type}:{job_id}"


def _prune_memory_store() -> None:
    now = time.time()
    expired = [k for k, exp in _memory_expiry.items() if exp < now]
    for k in expired:
        _memory_store.pop(k, None)
        _memory_expiry.pop(k, None)


async def _write_raw(job_type: str, job_id: str, value: dict[str, Any], ttl_seconds: int) -> None:
    k = _key(job_type, job_id)
    if is_redis_configured():
        await get_redis_client().set(k, json.dumps(value), ex=ttl_seconds)
        return
    _prune_memory_store()
    _memory_store[k] = value
    _memory_expiry[k] = time.time() + ttl_seconds


async def _read_raw(job_type: str, job_id: str) -> Optional[dict[str, Any]]:
    k = _key(job_type, job_id)
    if is_redis_configured():
        raw = await get_redis_client().get(k)
        if raw is None:
            return None
        return raw if isinstance(raw, dict) else json.loads(raw)
    _prune_memory_store()
    return _memory_store.get(k)


def _recover_if_stale(job_type: str, job_id: str, job: dict[str, Any]) -> dict[str, Any]:
    if job.get("status") == "running" and job.get("epoch") != PROCESS_EPOCH:
        corrected = {
            **job,
            "status": "failed",
            "error": "This job was still running when the AI service restarted and could not be resumed. Please retry.",
            "recovered": True,
        }
        # Best-effort persist of the correction — don't block the read on it.
        asyncio.create_task(_write_raw(job_type, job_id, corrected, DEFAULT_TTL_SECONDS))
        return corrected
    return job


async def create_job(job_type: str, job_id: Optional[str] = None, ttl_seconds: int = DEFAULT_TTL_SECONDS) -> str:
    """Creates a new 'running' job. Returns the jobId (generated if not supplied)."""
    jid = job_id or str(uuid.uuid4())
    job = {"status": "running", "epoch": PROCESS_EPOCH, "startedAt": time.time()}
    await _write_raw(job_type, jid, job, ttl_seconds)
    return jid


async def get_job(job_type: str, job_id: str) -> Optional[dict[str, Any]]:
    job = await _read_raw(job_type, job_id)
    if job is None:
        return None
    return _recover_if_stale(job_type, job_id, job)


async def update_job(
    job_type: str, job_id: str, patch: dict[str, Any], ttl_seconds: int = DEFAULT_TTL_SECONDS
) -> Optional[dict[str, Any]]:
    existing = await _read_raw(job_type, job_id)
    if existing is None:
        return None
    updated = {**existing, **patch}
    await _write_raw(job_type, job_id, updated, ttl_seconds)
    return updated
